core/src/main/scala/kafka/server/RaftServer.scala

```diff
@@ -101,30 +106,39 @@ class RaftServer(val config: KafkaConfig) extends Logging {
 
     socketServer.startProcessingRequests(Map.empty)
 
-    val shutdown = new AtomicBoolean(false)
-    try {
-      val counter = new ReplicatedCounter(config.brokerId, logContext, true)
-      val incrementThread = new Thread() {
-        override def run(): Unit = {
-          while (!shutdown.get()) {
-            if (counter.isLeader) {
-              counter.increment()
-              Thread.sleep(500)
-            }
-          }
-        }
-      }
+    val counter = new ReplicatedCounter(config.brokerId, logContext, true)
+    raftClient.initialize(counter)
 
-      raftClient.initialize(counter)
-      incrementThread.start()
+    raftIoThread = new RaftIoThread(raftClient)
+    raftIoThread.start()
 
-      while (true) {
-        raftClient.poll()
-      }
-    } finally {
-      shutdown.set(true)
-      raftClient.shutdown(5000)
-    }
+    incrementCounterThread = new IncrementCounterThread(counter)
+    incrementCounterThread.start()
+  }
+
+  def shutdown(): Unit = {
+    if (incrementCounterThread != null)
+      CoreUtils.swallow(incrementCounterThread.shutdown(), this)
+    if (raftIoThread != null)
+      CoreUtils.swallow(raftIoThread.shutdown(), this)
+    if (dataPlaneRequestHandlerPool != null)
+      CoreUtils.swallow(dataPlaneRequestHandlerPool.shutdown(), this)
+    if (socketServer != null)
+      CoreUtils.swallow(socketServer.shutdown(), this)
+    if (networkChannel != null)
+      CoreUtils.swallow(networkChannel.close(), this)
+    if (scheduler != null)
+      CoreUtils.swallow(scheduler.shutdown(), this)
+    if (metrics != null)
+      CoreUtils.swallow(metrics.close(), this)
+    if (metadataLog != null)
```

Comment:
In KafkaServer we first stopped the log-manager (including all the logs) and then the socket server, and I'm wondering if we should do the same -- although for simulation purposes it does not matter since we would not expect more requests but in general it seems a better pattern to me.

Reply:
Hmm, why do you think that is better? Stopping the socket server prevents additional requests. Seems like a good idea to do that before shutting down other internal components.
From the perspective of the Raft server, the main thing that matters is that the server continues handling requests until graceful shutdown has a chance to complete.

Reply:
Should we create a ticket to discuss KafkaServer shutdown ordering then?

Reply:
I believe there are issues already. But the order for the kafka broker is a little more subtle in general because there are other aspects like deregistration from zookeeper.